torch/quantization/observer.py

```diff
@@ -103,6 +103,127 @@ def calculate_qparams(self, **kwargs):
         return self._calculate_qparams(self.min_val.item(), self.max_val.item())
 
 
+class HistogramObserver(ObserverBase):
+    r"""
+    The module records the running histogram of tensor values along with
+    min/max values. calculate_qparams will calculate scale and zero_point
+    """
+
+    def __init__(self, bins=2048, **kwargs):
+        super(HistogramObserver, self).__init__(**kwargs)
+        self.bins = bins
+        self.histogram = None
+        self.min_val = None
+        self.max_val = None
+
+    def _combine_histograms(
+        self, dst_histogram, dst_min, dst_max, src_histogram, src_min, src_max
+    ):
+        bins_dst = dst_histogram.size()[0]
+        bins_src = src_histogram.size()[0]
+
+        dst_bin_width = (dst_max - dst_min) / bins_dst
+        src_bin_width = (src_max - src_min) / bins_src
+
+        for i in range(bins_src):
+            src_bin_count = src_histogram[i].item()
+            if src_bin_count == 0:
+                continue
+
+            src_bin_begin = src_min + src_bin_width * i
+            src_bin_end = src_bin_begin + src_bin_width
+
+            dst_bin = 0
+            if dst_bin_width:
+                dst_bin = int((src_bin_begin - dst_min) / dst_bin_width)
+
+            dst_bin_begin = dst_min + dst_bin_width * dst_bin
+            dst_bin_end = dst_bin_begin + dst_bin_width
+
+            dst_bin2 = 0
+            if dst_bin_width:
+                dst_bin2 = min(
+                    int((src_bin_end - dst_min) / dst_bin_width), bins_dst - 1
+                )
+
+            assert dst_bin2 <= dst_bin + 2, "1 src_bin is mapped to at most 2 dst_bins"
+            # dst_bin_cnt is the count from src_bin that should go to dst_bin
+            # the remainder should go to dst_bin2
+            dst_bin_cnt = 0
+            if src_bin_width == 0 or dst_bin_width == 0:
+                dst_bin_cnt = src_bin_count
+            else:
+                # We divide counts in src_bin in proportion to range overlap with dst_bin
+                dst_bin_cnt = min(
+                    round(
+                        (dst_bin_end - src_bin_begin) / src_bin_width * src_bin_count
+                    ),
+                    src_bin_count,
+                )
+
+            dst_histogram[dst_bin] += dst_bin_cnt
+
+            # remaining should go to dst_bin2
+            if dst_bin_cnt < src_bin_count:
+                dst_histogram[dst_bin2] += src_bin_count - dst_bin_cnt
+
+    def forward(self, x):
+        if self.min_val is None or self.max_val is None or self.histogram is None:
+            self.min_val = torch.min(x)
+            self.max_val = torch.max(x)
+            self.histogram = torch.histc(x, self.bins)
+        else:
+            new_min = torch.min(x)
+            new_max = torch.max(x)
+            new_histogram = torch.histc(x, self.bins)
+            # combine the existing histogram and new histogram into 1 histogram
```

Comment:
For the first version of histogram observer, suggest having a fixed width histogram, so that combining histograms is trivial.
```python
 def forward(self, x):
        if self.min_val is None or self.max_val is None or self.histogram is None:
            self.min_val = torch.min(x)
            self.max_val = torch.max(x)
            range = self.max_val -self.min_val
            self.relaxed_min = self.min_val - 0.5*range
            self.relaxed_max = self.max_val + 0.5*range
            self.histogram = torch.histc(input, self.bins, min=self.relaxed_min, max=self.relaxed_max)
        else:
            new_min = torch.min(x)
            new_max = torch.max(x)
            new_histogram = torch.histc(x, self.bins, min=relaxed_min, max=relaxed_max)
            self. histogram = new_histogram+self.histogram
```